### ISSUE TITLE
Crash in PTVS 2.1 when saving options

### ISSUE THREAD

**Issue by dpacbach** [Microsoft/PTVS#937]
Hi, I'm using the latest version of PTVS 2.1 branch that I've built with VS2013 and it works fine in general, but I am sometimes seeing crashes (or freezes) upon hitting OK in the environment options dialog.  The line that throws the uncaught exception is:

``` csharp
var replProvider = PythonToolsPackage.ComponentModel.GetService<IReplWindowProvider>();
```

in the PythonInteractiveOptionsPage.cs file.  Does anyone know what this is and/or has there been a fix for it?  I remember having this problem last time I tried v2.1, then I upgraded to a more recent version and I think the problem went away.  But unfortunately I now need to use version 2.1 for various reasons and am seeing this problem.  Is there a fix or workaround?  Thank you

**Comment by zooba** [comment]
Do your Interactive Windows work okay? Does it make a difference to this hang if you've already opened some vs. never opening an interactive window?

**Comment by dpacbach** [comment]
Hi zooba,
Thanks for the response -- I reset my Exp hive and now this issue seems to have gone away, but not sure why.  But now I'm having another problem, and that is that when I create a new python project from the New Project window (which seems to work fine), then I right click on the project name in Solution Explorer I get an uncaught exception.  The same thing happens if I select the project in Solution explorer and then hover my mouse over the PROJECT menu (basically I'm just trying to edit the project settings but it keeps crashing).  Any idea what this could be and if it's related to the first problem?  Thank you,
David

**Comment by dpacbach** [comment]
Hi zooba,
Here are some more details on what I did/am observing:
  1) Check out the head of the 2.1 branch
  2) Do a "minimal" build, which succeeds
  3) Deploy to Exp for debugging (although the problem happens whether in debug mode or not)
  4) Create a new project from one of the project templates (which succeeds in the Exp hive)
  5) Right click on the project in solution explorer in order to change some properties
  6) The crash happens immediately upon right clicking, before the context menu pops up.  It also crashes when repeating steps 1-4 but then clicking on the PROJECT menu

Currently I only have the two default 2.6/2.7 python environments set up which work fine -- I can open a REPL without a problem.

Over the last few months I've been doing most of my testing with version 2.2 I believe, where I did not have these problems.  But recently I found out that I need to use version 2.1 and now these issues are popping up (crashing when right-clicking on the project and also sometimes crashing when saving options).  Any ideas?

Thank you,
David

**Comment by zooba** [comment]
(As an aside, I'm interested in why you need 2.1 and not 2.2. You can email our team privately at ptvshelp@microsoft.com if you don't want to explain publicly.)

So we actually dropped VS 2013 from 2.1.1, which is at the tip of 2.1, and it technically only supports VS 2010 and 2012. This should only matter for one bit of Django support, which I removed rather than attempting to handle all the edge cases involved in VS 2013 (see `TemplateArtifactCollection.cs` if you're interested), but it shouldn't affect the REPL window.

What's more likely is things failing to load correctly because of the non-standard deployment. Check whether you are deploying `Microsoft.VisualStudio.ReplWindow.dll` into the Python Tools extension directory. It is supposed to only be in the ReplWindow extension, but it is also supposed to be in the GAC which may have been masking some other resolution issues.

If you debug the VS instance, break on all exceptions and then cause the crash you may be able to extract some more information about the source. These exceptions often include `InnerException` or `LoaderExceptions` properties that are more informative than the one that is actually thrown.

**Comment by dpacbach** [comment]
Hi zooba,
I did a release build of the extension, then took the vsix/dll files and installed them into another hive using a vsix installation program and then the issue seems to go away, so looks like you're right that it is deployment-related.

That said, it would be nice to be able to use the projects feature within the Exp hive while debugging -- this is because I haven't found a way to enter into mixed mode debugging without running in the context of a project and checking the "Enable native code debugging" checkbox... and similarly I haven't found a way to add folders into the PYTHONPATH of the interpreter without using a project... am I missing something?

Ideally actually I wouldn't create any projects at all for most use cases, I would just open a new python file, type some code, then debug it in mixed-mode and have it automatically jump to the C++ code that is being called.  Is it possible to do that without a project?

Regarding the version issue, I work at a large firm that is slow to approve new software (currently v2.1 is approved but not yet 2.2).  PTVS is just starting to be introduced on a large scale here and so I'm sure we'll upgrade eventually, but for now have to stick to 2.1.

**Comment by zooba** [comment]
You're right that those features are currently unavailable without projects, though we are hoping to fix that for 3.0. Given Visual Studio is built around the concept of projects, we really have no choice but to support them, and reimplementing everything twice to cover non-project scenarios is a really big ask.

If you're getting PTVS introduced widely then we'd *definitely* like to hear more information - the more users we have the easier it is to get more developers to work on PTVS. Also, PTVS is now available as part of VS 2015, so if that's being approved then it should include PTVS 2.2 (depending on how thorough your firm is).

**Comment by dpacbach** [comment]
Hi zooba, I will send an email to steve.dower@microsoft.com if you don't mind.  -David
